### PR TITLE
More thread-safe GC

### PR DESCRIPTION
**Pull request by cjdoris**
A draft of an alternative to #520 to ensure Python GC only runs on thread 1.

This version avoids using tasks. Instead, when a Python object is finalized:
* If GC is enabled and the GIL is currently held, decref the object immediately.
* Otherwise add it to the queue.

I also added `PythonCall.GC.gc()` to explicitly clear the queue (if possible).

Furthermore I added an immortal object `GCHook()` which calls `PythonCall.GC.gc()` whenever it is finalized, then adds its own finalizer back on. This means the queue can be cleared periodically when Julia GC runs, because it keeps trying to finalize `GCHook()`.

I used `PyGILState_Check() == 1` to check that the GIL is currently held. This should make it safe to interact with Python from any thread in the future, provided you use the C API `PyGILState_*` and `PyThreadState_*` functions correctly.

Very much a draft. Need to use `Channel`s etc (like #520) to make this totally thread-safe.

### CODE REVIEW

**Comment by ericphanson** [comment]
BTW, could be worth trying this benchmark: https://github.com/JuliaPy/PythonCall.jl/compare/main...ericphanson:PythonCall.jl:eph/timing

**Comment by cjdoris** [comment]
@ericphanson @lassepe @MilesCranmer I think this is the version I'm going to merge.

I'd be grateful if you could give it a try to check it solves any issues, and also review the code.

Some timings on my machine:
* `pyint(123456789)` (quite a fast operation) takes 45ns on the main branch versus 54ns on this branch if the GIL is held.
* That slow-down is expected, because the main difference is the addition of a `C.PyGILState_Check()` call, which is 11ns.
* If the GIL is not held when the finalizer is called then the finalizer takes an extra 20ns, half being from locking/unlocking the spin lock, the other half being pushing/popping the queue. On main you get a slowdown of 10ns if GC is disabled (from pushing/popping), so this PR adds 10ns (locking/unlocking) in this situation.
* Using channels instead of vectors and locks adds about 80ns to the no-GIL branch. So I don't use channels.
* The fastest you can possibly create and free this int (`C.Py_DecRef(C.PyLong_FromLongLong(123456789))`) takes 19ns.
* In short, we were 2.25x slower than optimal before and now we are 2.75x slower. Or another way, we could make 22.2M ints per second before, now we can make 18.5M. This seems acceptable to me for the improvement.

I just need to copy over the extra tests from #520.

**Comment by MilesCranmer** [comment]
PySR test suite is passing :+1:

**Comment by MilesCranmer** [comment]
Cool, this also fixes https://github.com/avik-pal/Wandb.jl/issues/27!

File: src/GC/GC.jl
Comment (MilesCranmer) [inline src/GC/GC.jl:13]: Nitpick is that I would prefer a pattern where these are one object so that it's more obvious to another developer that there is a lock associated to this queue, like:

```julia
const QUEUE = (;
    x=C.PyPtr[],
    lock=Threads.SpinLock()
)
```

and then you access it with

```julia
lock(QUEUE.lock) do
    f(QUEUE.x)
end
```

To me this is a more obvious association than the name + `_LOCK`. If a future dev does a global rename of `QUEUE`, they might forget to update the associated lock's name, and they would dissociate.

File: src/GC/GC.jl
Comment (MilesCranmer) [inline src/GC/GC.jl:109]: Just wondering how/if this interacts with Julia multiprocessing and Python multiprocessing, if at all?

**Review by MilesCranmer — COMMENTED** [review]
I think the docs also need to be updated with the warning about multithreading (and suggestion that `GC` be turned off).

Also, how does this interact with the signal handling, does the warning in the docs still apply?

File: src/GC/GC.jl
Comment (MilesCranmer) [inline src/GC/GC.jl:26]:
```suggestion
disable() = (Base.depwarn("disabling the PythonCall GC is no longer needed", :disable); nothing)
```

File: src/GC/GC.jl
Comment (MilesCranmer) [inline src/GC/GC.jl:38]:
```suggestion
enable() = (Base.depwarn("disabling the PythonCall GC is no longer needed", :enable); nothing)
```

File: src/GC/GC.jl
Comment (MilesCranmer) [inline src/GC/GC.jl:87]:
```suggestion
    if any(Base.Fix2(!=, C.PYNULL), ptrs) && C.CTX.is_initialized
```
(reduces compilation load)

File: src/GC/GC.jl
Comment (MilesCranmer) [inline src/GC/GC.jl:59]: I prefer the `lock(...) do ...` API since it will still unlock in the case of an error.

File: src/GC/GC.jl
Comment (MilesCranmer) [inline src/GC/GC.jl:98]: Prefer `lock(...) do ...`

File: src/GC/GC.jl
Comment (MilesCranmer) [inline src/GC/GC.jl:58]: Why is it `unsafe_` if it is behind the lock? I would have assumed the `unsafe_` code would need to be locked from outside.

File: src/GC/GC.jl
Comment (MilesCranmer) [inline src/GC/GC.jl:6]: I'm assuming these can be removed from the module docstring since no longer in use?

File: src/GC/GC.jl
Comment (MilesCranmer) [inline src/GC/GC.jl:116]: Just curious, why `mutable`? (Maybe there's a deeper reason for interacting with Julia GC?)

File: src/GC/GC.jl
Comment (cjdoris) [reply src/GC/GC.jl:25]: `Base.depwarn` is not documented and there doesn't seem to be an equivalent in the API

File: src/GC/GC.jl
Comment (cjdoris) [reply src/GC/GC.jl:37]: As above, `Base.depwarn` is not API sadly

File: src/GC/GC.jl
Comment (cjdoris) [reply src/GC/GC.jl]: Yeah I considered it but (a) it adds 30ns and (b) the code in between should not throw.

File: src/GC/GC.jl
Comment (cjdoris) [reply src/GC/GC.jl:57]: See `gc()` - you need to check that `C.CTX.is_initialized` is true.

**Comment by cjdoris** [comment]
> I think the docs also need to be updated with the warning about multithreading (and suggestion that `GC` be turned off).
> 
> Also, how does this interact with the signal handling, does the warning in the docs still apply?

Thanks I have updated the docs.

The signal handling warning still applies. #520 enabled it by default when multi-threading is enabled but I'm not changing it in this PR. Happy to discuss it though.

**Comment by cjdoris** [comment]
I have copied the tests across from #520 and they're passing so I think this is ready to merge.

Will give you all a chance to comment.

**Comment by cjdoris** [comment]
BTW this PR makes it possible to write code that calls into Python from multiple threads (or Python code calling into Julia from multiple threads), provided you handle the GIL correctly. I'm planning to add an API for safely acquiring and pausing the GIL after this PR.

File: src/GC/GC.jl
Comment (MilesCranmer) [inline src/GC/GC.jl:25]:
```suggestion
@deprecate disable() identity(nothing) false
```

This would be the documented equivalent. But normally if there's no alternative function to call you just write a message manually with `Base.depwarn`: https://github.com/search?q=/Base.depwarn/%20language:julia&type=code

File: src/GC/GC.jl
Comment (MilesCranmer) [inline src/GC/GC.jl:37]:
```suggestion
@deprecate enable() identity(nothing) false
```

File: src/GC/GC.jl
Comment (MilesCranmer) [reply src/GC/GC.jl:25]: (It's helpful for library maintainers, since the `depwarn` will show up during unittests but not during normal execution – so I know what functions I need to update.)

Comment (MilesCranmer) [reply src/GC/GC.jl:25]: Since if you do it with the `@deprecated` API, it doesn't really make sense in this context:

```julia
julia> function foo()
       end
foo (generic function with 1 method)

julia> function _foo()
       end
_foo (generic function with 1 method)

julia> @deprecate foo() _foo()
foo (generic function with 1 method)

julia> foo()
┌ Warning: `foo()` is deprecated, use `_foo()` instead.
│   caller = top-level scope at REPL[4]:1
└ @ Core REPL[4]:1
```

File: src/GC/GC.jl
Comment (MilesCranmer) [reply src/GC/GC.jl:37]: Though again, I'd prefer the `Base.depwarn`, otherwise it's a bit confusing to package maintainers

File: src/GC/GC.jl
Comment (MilesCranmer) [reply src/GC/GC.jl]: I guess it's the additional function call from the closure. In that case I would try

```julia
lock(QUEUE_LOCK)
try
    #= do stuff =#
finally
    unlock(QUEUE_LOCK)
end
```
And hopefully that's a bit faster.

If anything it's just a good habit. I think there's always a possibility for some kind of error in Julia code until we can actually impose such constraints on code, so locks should always be safely handled. e.g., perhaps the user could `<Ctrl-C>` while the lock is in place, or there's an OOM error, or something else we haven't thought of.

File: src/GC/GC.jl
Comment (MilesCranmer) [inline src/GC/GC.jl]:
```suggestion
            lock(QUEUE_LOCK)
            try
                push!(QUEUE, ptr)
            finally
                unlock(QUEUE_LOCK)
            end
```

File: src/GC/GC.jl
Comment (MilesCranmer) [reply src/GC/GC.jl]: (As above)

**Comment by MilesCranmer** [comment]
Regarding `Base.depwarn`, it looks like it was just officially documented and declared part of the public API: https://github.com/JuliaLang/julia/blame/19165bec4a3e2ed0de4f2f43bccc3fcac2d76662/base/deprecated.jl#L230-L249

File: src/GC/GC.jl
Comment (ericphanson) [reply src/GC/GC.jl]: BTW `Base.@lock` (which is what I used in my PR) expands to exactly the code you wrote there @MilesCranmer, and is exported on more recent julia versions

File: src/GC/GC.jl
Comment (ericphanson) [inline src/GC/GC.jl:12]: Base has `Lockable` starting in 1.11 (exported in 1.12) FYI. So not useful here but just so you know :). It is just like this, except it also enforces the lock is held when you access the item it holds via indexing.

File: src/GC/GC.jl
Comment (MilesCranmer) [reply src/GC/GC.jl]: Note this is also the same pattern recommended by the docs: https://docs.julialang.org/en/v1/base/parallel/#Base.ReentrantLock as well as the pattern that is generated by the `Base.@lock` macro: https://docs.julialang.org/en/v1.11.0-rc2/base/parallel/#Base.@lock

File: src/GC/GC.jl
Comment (MilesCranmer) [reply src/GC/GC.jl:25]: Unresolving this

File: src/GC/GC.jl
Comment (MilesCranmer) [inline src/GC/GC.jl:25]:
```suggestion
disable() = (Base.depwarn("disabling the PythonCall GC is no longer needed", :disable); nothing)
```
Unresolving this as we've identified the docs and this function being official public API: https://github.com/JuliaLang/julia/pull/55212

File: src/GC/GC.jl
Comment (MilesCranmer) [inline src/GC/GC.jl:37]:
```suggestion
enable() = (Base.depwarn("disabling the PythonCall GC is no longer needed", :disable); nothing)
```

Unresolving this as we've identified the docs and this function being official public API: https://github.com/JuliaLang/julia/pull/55212

**Comment by MilesCranmer** [comment]
I think apart from putting the lock behind a `try`-`finally`, it's otherwise looking good.

My one other comment is more of a question (for both you and @ericphanson) – why `Threads.SpinLock()` and not `ReentrantLock()`?

**Comment by ericphanson** [comment]
I think it is because we can't yield to other tasks in finalizers so we have to busywait: 

> f must not cause a task switch, which excludes most I/O operations such as println. Using the @async macro (to defer context switching to outside of the finalizer) or ccall to directly invoke IO functions in C may be helpful for debugging purposes.


from https://docs.julialang.org/en/v1/base/base/#Base.finalizer

**Comment by MilesCranmer** [comment]
I think I'm misreading, because a reentrant lock wouldn't allow that, right? Reentrant locks are about letting the same task acquire the lock multiple times?

File: src/GC/GC.jl
Comment (MilesCranmer) [inline src/GC/GC.jl:103]: Tangentially related, could each of these functions have a few more inline comments? Just to help me figure out the purpose of each block.

**Comment by ericphanson** [comment]
I edited in a response a bit late sorry. ReentrantLock vs SpinLock is not only about reentry but also implementation, since task-switching is an implementation detail. ReentrantLocks do it and SpinLocks don't.

See: [ReentrantLock waiting](https://github.com/JuliaLang/julia/blob/19165bec4a3e2ed0de4f2f43bccc3fcac2d76662/base/lock.jl#L157)

[SpinLock implementation](https://github.com/JuliaLang/julia/blob/19165bec4a3e2ed0de4f2f43bccc3fcac2d76662/base/locks-mt.jl#L41-L71)

File: src/GC/GC.jl
Comment (ericphanson) [reply src/GC/GC.jl:103]: It won't let me suggest comments on deleted lines, but my attempt:

* `any(!=(C.PYNULL), ptrs)`: is there any work to do? if the pointers are all null then nothing to do
* `C.CTX.is_initialized`: need to ensure the context is initialized (not sure what that does?)
* `C.PyGILState_Check() == 1`: does this thread hold the GIL?
* if so, eagerly decrease the reference count on each object without putting it in the queue. Then since we know we hold the GIL, for good measure also process the queue built up on other threads
* if not, then acquire the queue lock so we can add these items to the queue

**Comment by MilesCranmer** [comment]
Maybe if we use the same pattern from https://github.com/JuliaPy/PyCall.jl/pull/1074 we can get a speedup? If we have a `lock(::SpinLock)`, the GC threads are forced to wait for each other. But if we do the equivalent of the following code from PyCall (never implemented; just a PR)

```julia
function _pydecref_locked(po::PyObject)
    # If available, we lock and decref
    !islocked(PYDECREF_LOCK) &&
        trylock(() -> pydecref_(po), PYDECREF_LOCK) &&
        return nothing

    # Otherwise, add back to queue to be decref'd later
    finalizer(_pydecref_locked, po)
    return nothing
end
```

then the threads never need to have a busy wait on the lock to be released, they just add it back to the queue and get on with some other work.

File: src/GC/GC.jl
Comment (ericphanson) [inline src/GC/GC.jl:132]: hm, I'm not sure this does what you want. The `WeakRef` constructor does a [ccall](https://github.com/JuliaLang/julia/blob/19165bec4a3e2ed0de4f2f43bccc3fcac2d76662/base/boot.jl#L451) and we are bypassing that.

What about `HOOK` being a Ref holding a weakref? 😅. Then we can construct the weakref via its constructor and stick it in `HOOK`.

**Comment by ericphanson** [comment]
That is https://github.com/JuliaPy/PythonCall.jl/pull/534, right? maybe that strategy should be discussed in that PR

**Comment by MilesCranmer** [comment]
Oh yeah good point:

> To me some kind of queue makes the most sense. However probably anything is better than the segfaults of the status quo.

I agree, I think we can just worry about further optimizations later.

File: src/GC/GC.jl
Comment (cjdoris) [reply src/GC/GC.jl]: The way I did it did appear to work. However assigning to `.value` is not documented behaviour so I have switched to your suggestion. We're only keeping the weak reference around for debugging purposes.

**Comment by ericphanson** [comment]
I think we need SpinLock since we are in a finalizer! We aren’t allowed to yield like ReentrantLock does

**Review by MilesCranmer — APPROVED** [review]
Looks great!

File: pytest/test_all.py
Comment (ericphanson) [inline pytest/test_all.py:84]: This comment is out of date (no background task, no printing)

**Review by ericphanson — APPROVED** [review]
One out of date comment, otherwise lgtm (though I should say I’m not expert and was just learning as I went in #520)

File: src/GC/GC.jl
Comment (MilesCranmer) [inline src/GC/GC.jl:142]: Does this `isempty` need to be locked?

Looking at the implementation: https://github.com/JuliaLang/julia/blob/48d4fd48430af58502699fdf3504b90589df3852/base/essentials.jl#L957 it might need to be.

File: src/GC/GC.jl
Comment (MilesCranmer) [inline src/GC/GC.jl:113]: i think the entire if statement should be locked, rather than just the call to `append!`. See my other comment — you might otherwise be trying to `iterate` through since-removed elements.

File: src/GC/GC.jl
Comment (MilesCranmer) [inline src/GC/GC.jl:90]: Similarly, the whole of statement should be locked.

Comment (MilesCranmer) [reply src/GC/GC.jl:90]: Or at least the `isempty` call by itself. I think it can segfault if another thread is making concurrent writes

File: src/GC/GC.jl
Comment (cjdoris) [reply src/GC/GC.jl:142]: No it's this method (just checking length == 0): https://github.com/JuliaLang/julia/blob/1dffd7752de2409b5f6c81fdcc1f33118127725e/base/abstractarray.jl#L1237

I believe this is safe. At worst, if there is a race condition, the `isempty` check might return `true` when it should be `false` or vice-versa, but all this means is that we might not empty the queue when we should, or we might waste time emptying an already-empty queue. But neither of these is really a problem - we'll just empty the queue next time around. It's suuuuuper unlikely such a race condition would strike many times in succession.

File: src/GC/GC.jl
Comment (cjdoris) [reply src/GC/GC.jl:113]: All mutation of the queue is behind the lock, so I think this is fine given my other comment about `isempty`.

File: src/GC/GC.jl
Comment (cjdoris) [reply src/GC/GC.jl:90]: (ditto my other comment)

**Comment by MilesCranmer** [comment]
Regarding the `isempty` needing to be locked,
```julia
iterate(items) === nothing
```
This is how Base checks for an array being empty. I think what can happen is: thread A can (1) call the `iterate`, (2) get the pointer to the first array element, and then (3) a separate thread B removes that element from the array, making that pointer point to invalid memory, (4) thread A accesses the first array element anyways.

In other words, use of `isempty` could cause undefined behavior if another thread is writing to the collection (and maybe a segfault).

**Comment by cjdoris** [comment]
> Regarding the `isempty` needing to be locked,

I [commented elsewhere](https://github.com/JuliaPy/PythonCall.jl/pull/529#discussion_r1702283046) that you're looking at the wrong method of `isempty`. [The method for arrays](https://github.com/JuliaLang/julia/blob/master/base/abstractarray.jl#L1237) is
```julia
isempty(a::AbstractArray) = (length(a) == 0)
```
which I believe is safe. It might give wrong answers occasionally in the presence of a race condition, but it's safe.

**Comment by MilesCranmer** [comment]
Oh yeah, you are totally right, sorry.

**Comment by MilesCranmer** [comment]
Ok, no other comments from me. I think it's ready to go :+1:

**Comment by cjdoris** [comment]
Thanks for your help reviewing!

**Comment by MilesCranmer** [comment]
No problem :+1:

(By the way, could we get a release so we can fix the Wandb.jl issue?)

**Comment by cjdoris** [comment]
Yeah I'll make a release once #535 is merged (going to sleep on it and bikeshed the new function names).

**Comment by MilesCranmer** [comment]
Sounds good.

In case #535 seems like it will take some more discussion to settle on an API I wonder if we could have a patch release just so we can get the bug fixes through?

**Comment by MilesCranmer** [comment]
@cjdoris could we have a release before #535? I need the thread safety for something PySR-related and it's slightly urgent (am teaching a summer school this week and want to have them using Wandb.jl)

**Comment by cjdoris** [comment]
It's released

**Comment by MilesCranmer** [comment]
Thanks!

**Pull request merged**
